rust/dyn-ddlog/src/optimize/expr.rs

```diff
@@ -0,0 +1,493 @@
+//! Query and expression optimization using [`egg`]
```

Comment:
Given that we compile to Rust, doesn't it make sense to rely on Rust to perform this type of optimization?

As far as expression translation goes, the real problem that requires smarts in the compiler is reducing cloning.  In DDlog all variables bind by value, which leads to simple expression semantics, but, when implemented naively, leads to redundant cloning. Mapping DDlog dataflow to efficient Rust would be nice.

Reply:
We don't always compile to rust and this is also important in conjunction to dataflow optimization as it can reduce complicated expressions into things we can really aggressively optimize, such as `.reduce(|coll| coll.iter().sum())` into `.explode(|key, val| (key, Sum::new(val))` (lifting the summed value into the difference type). Additionally once things get more complex egg can be exactly what does the clone reduction optimizations since we can express really complex things like ownership within it. As the `optimize` module (I think) says, it's all experimental so don't worry about it a ton

Reply:
> We don't always compile to rust

ok, good point

> Additionally once things get more complex egg can be exactly what does the clone reduction optimizations since we can express really complex things like ownership within it.

Right. I've no idea how to model this sort of dataflow analysis in egg, but I am looking forward to finding out.

> As the optimize module (I think) says, it's all experimental so don't worry about it a ton

Sure, I'm just getting up to speed with the code base.